### DISCUSSION TITLE
Relationships in the Database

### DISCUSSION THREAD

**Discussion by adityaoberai** [appwrite/appwrite#5179] [Ideas]
Hey folks 👋 

Lately, we have been working on one of the most requested features to improve our Databases service: Relationships 🥳

We've created an RFC for database relationships to explain the purpose and implementation of this feature: [Database Relationships - RFC](https://github.com/appwrite/rfc/blob/0ab31760035f5c312a89aafd281a7e97370ffd72/021-%20database-relationships.md)

Additionally, this feature requires some changes in the Appwrite Console. To see how this new feature might work, we are inviting you to try the Figma prototype below. The prototype simulates the creation of a **many-to-one relationship** between a collection called `Review` to a collection called `Movie`.

https://www.figma.com/proto/WxczFR7RG5OLKMnC0EW9YD/Databases?node-id=2283%3A140215&scaling=scale-down&page-id=2283%3A136756&starting-point-node-id=2283%3A140215

We would love for you folks to have a look at the RFC and the prototype and share your thoughts and feedback here 🙏

**Comment by squallsama** [comment]
What about cascade delete when parent doc removed ? For example if Movie will be removed - Appwrite should remove `Reviews` associated with this Movie or at least give this option to set this in attributes.

**Reply by abnegate** [reply]
@squallsama Yep this will be an option for the relationship attribute

**Reply by Sashika-Dilmina** [reply]
Appwrite **does not support cascade delete**. If you delete a Movie, **its Reviews are not deleted automatically**. You must **manually delete Reviews** using a Cloud Function or in your code.

**Reply by adityaoberai** [reply]
> Appwrite **does not support cascade delete**. If you delete a Movie, **its Reviews are not deleted automatically**. You must **manually delete Reviews** using a Cloud Function or in your code.
> 

Actually, as @abnegate specified, we do support cascade deletes. This is specified in the docs too: https://appwrite.io/docs/products/databases/relationships#on-delete

**Comment by rhengles** [comment]
Who is `self::` here ?

> Creating a relationship with an SDK:
>
> ```php
> $client = new Client();
> $databases = new Databases(client);
>
> $relationship = $databases->createRelationship(
>     databaseId: 'music',
>     collectionId: 'albums',
>     relatedCollectionId: 'artists',
>     type: Relationship.ManyToOne,
>     key: 'artist',
>     twoWay: true,
>     twoWayId: 'albums',
>     onUpdate: self::CASCADE,
>     onDelete: self::RESTRICT
> );
> ```

**Reply by abnegate** [reply]
These should be `Database::CASCADE` and `Database::RESTRICT`, they are just constants for `'cascade'` and `'restrict'`. Will update :grin:

**Reply by Sashika-Dilmina** [reply]
self:: means you're using something (like a constant) from inside the same class.

But in your example, CASCADE and RESTRICT come from the Relationship class, not your own class

**Comment by squallsama** [comment]
Also would be great to have Row Level Security with ability to use allowed `userIds` from different collections

**Reply by stnguyen90** [reply]
What do you mean by this? Our permission system should already support row level security 🧐

**Reply by squallsama** [reply]
You can specify user ids in permission section, but you can't use queries inside Permission object.

**Reply by stnguyen90** [reply]
@squallsama, I see. That's quite unrelated to this topic. I would suggest creating an issue to request this type of feature.

**Reply by Mahaveerkirshanani** [reply]
**what** i did not understand**
@stnguyen90 [https://apkanamaekya.com](url)

**Comment by maciej-sz** [comment]
Are you also planning to support the relationships in nested queries of the GraphQL API?

**Reply by Sashika-Dilmina** [reply]
Appwrite’s Graph API does not support nested queries for relationships. This means you can’t fetch related data

**Comment by rafagazani** [comment]
@abnegate Will the Relationship type attributes have some protection to prevent the sdk client from updating these fields?

**Comment by JamieBradders** [comment]
Just wanted to say that I love the proposed UX (Figma Designs). Really can't wait to see this feature. Well done team and thank you! 🎉

**Comment by eldadfux** [comment]
Everyone, we're excited to share database relationships are now available as a beta feature in Appwrite 1.3!
You can learn more on our full announcement: https://dev.to/appwrite/join-celebrations-appwrite-13-ships-relationships-57fc

We're excited to here what you all think!

**Reply by D-Raj-Grg** [reply]
On the new version, 

Only 3rd level of relation data is shown. (listDocuments from 1st)


I'll get all four level data if I listDocuments from 2nd Level.

**Can we do one more level **4th?********

**Reply by hortigado** [reply]
Same i need 4 levels or more

**Comment by ikanindustries** [comment]
Is there a way to upgrade my appwrite cloud version?
